stdlib/public/runtime/Paths.cpp

```diff
@@ -0,0 +1,196 @@
+  return ch == '/' || ch == '\\';
+#else
+  return ch == '/';
+#endif
+}
+
+#if !defined(_WIN32) || defined(__CYGWIN__)
+#define PATHSEP_STR "/"
+#define PATHSEP_CHR '/'
+#else
+#define PATHSEP_STR "\\"
+#define PATHSEP_CHR '\\'
+#endif
+
+#ifdef _WIN32
+#define EXE_EXT ".exe"
+#else
+#define EXE_EXT ""
+#endif
+
+}
+
+SWIFT_RUNTIME_EXPORT
+const char *
+swift_getRuntimePath()
+{
+  swift::once(runtimePathToken, _swift_initRuntimePath, nullptr);
+  return runtimePath;
+}
+
+SWIFT_RUNTIME_EXPORT
+const char *
+swift_getRootPath()
+{
+  swift::once(rootPathToken, _swift_initRootPath, nullptr);
+  return rootPath;
+}
+
+namespace {
+
+void
+_swift_initRootPath(void *)
+{
+  // SWIFT_ROOT overrides the path returned by this function
+  char *swiftRoot = getenv("SWIFT_ROOT");
+  if (swiftRoot) {
+    size_t len = std::strlen(swiftRoot);
+    if (len) {
+      // Make sure it has a trailing path separator
+      if (_swift_isPathSep(swiftRoot[len - 1]))
+        rootPath = ::strdup(swiftRoot);
+      else {
+        char *path = (char *)std::malloc(len + 2);
+        std::memcpy(path, swiftRoot, len);
+        std::memcpy(path + len, PATHSEP_STR, 2);
+        rootPath = path;
+      }
+      return;
+    }
+  }
+
+  // Otherwise, get the runtime path and remove /lib/libswiftCore.(dylib|dll|so)
+  const char *runtimePath = swift_getRuntimePath();
+  const char *end = runtimePath + std::strlen(runtimePath);
+  const char *ptr = end;
+  int count = 0;
+
+  while (ptr > runtimePath) {
+    int ch = *--ptr;
+    if (_swift_isPathSep(ch)) {
+      if (++count == 2)
+        break;
+    }
+  }
+
+  if (count != 2) {
+    swift::fatalError(/* flags = */ 0,
+                      "libswiftCore should be installed in a `lib` "
+                      "directory: %s\n",
+                      runtimePath);
+  }
+
+  // Copy the path including the trailing '/'
+  size_t len = ptr - runtimePath + 1;
+  char *path = (char *)std::malloc(len + 1);
+  std::memcpy(path, runtimePath, len);
+  path[len] = 0;
+
+  rootPath = path;
+}
+
+}
+
+SWIFT_RUNTIME_EXPORT
+const char *
+swift_getAuxiliaryExecutablePath(const char *name)
+{
+  const char *rootPath = swift_getRootPath();
+
+  // Allocate a buffer for the result
+  const size_t nameLen = std::strlen(name);
+  const size_t extLen = std::strlen(EXE_EXT);
+  const size_t rootPathLen = std::strlen(rootPath);
+  const size_t libexecLen = 8;
+  char *result = (char *)std::malloc(rootPathLen + libexecLen
+                                     + nameLen + extLen + 1);
+  char *ptr = result;
+
+  std::memcpy(ptr, rootPath, rootPathLen); ptr += rootPathLen;
+  std::memcpy(ptr, "libexec" PATHSEP_STR, libexecLen); ptr += libexecLen;
+  std::memcpy(ptr, name, nameLen); ptr += nameLen;
+  std::memcpy(ptr, EXE_EXT, extLen + 1);
+
+  return result;
+}
+
+namespace {
+
+#if !defined(_WIN32) || defined(__CYGWIN__)
+void
+_swift_initRuntimePath(void *) {
+  Dl_info dli;
+
+  int ret = ::dladdr((void *)_swift_initRuntimePath, &dli);
```

Comment:
This should use the SPI `dyld_image_path_containing_address` where available, as it's slightly more efficient. There's precedent at https://github.com/apple/swift/blob/0213f2dd884ec4286d3aee03528cbfc6a961d54c/stdlib/public/runtime/ObjCRuntimeGetImageNameFromClass.mm#L82